frame/support/procedural/src/pallet/expand/call.rs

```diff
@@ -54,6 +54,35 @@ pub fn expand_call(def: &mut Def) -> proc_macro2::TokenStream {
 		.map(|fn_name| format!("Create a call with the variant `{}`.", fn_name))
 		.collect::<Vec<_>>();
 
+	let mut warning_structs = Vec::new();
+	let mut warning_names = Vec::new();
+	// Emit a warning for each call that is missing `call_index` when not in dev-mode.
+	for method in &methods {
+		let explicit_call_index =
+			crate::pallet::parse::helper::take_item_pallet_attrs(&mut method.clone().attrs)
+				.unwrap_or_default()
+				.into_iter()
+				.any(|attr| matches!(attr, crate::pallet::parse::call::FunctionAttr::Weight(_)));
+		if explicit_call_index || def.dev_mode {
+			continue
+		}
+
+		let name =
+			syn::Ident::new(&format!("implicit_call_index_{}", method.name), method.name.span());
+		let warning: syn::ItemStruct = syn::parse_quote!(
+			#[deprecated(note = r"
+			`pallet::call_index` is missing on a call while the dev-mode is disabled.
+			Please ensure that either all your calls have the `pallet::call_index` attribute or
+			that the dev-mode for your pallet is enabled. For more info see:
+			<https://github.com/paritytech/substrate/pull/11381> and
+			<https://github.com/paritytech/substrate/pull/12536>.")]
```

Comment:
Huh... couldn't we just have put this attribute right on top of the Call enum variant or something similar?

Reply:
Yea, was not sure how to best approach this. But having it in a dedicated struct allows for more warnings in the future.  
Or do you think it will not correctly work like this?

Reply:
Okay I tried that but it has two down-sides:  
- I cannot add a span anymore, so the generated links just point to the expanding `#[frame_support::pallet]` instead of the line of the call.
- It now reports one warning per call argument, since apparently deprecating an enum variant also deprecates all its fields:  

1.
``pre
warning: use of deprecated field `pallet::Call::refund::id`: 
                     Implicit call indices are deprecate in favour of explicit ones.
                     Please ensure that all calls have the `pallet::call_index` attribute or that the
                     `dev-mode` of the pallet is enabled. For more info see:
                     <https://github.com/paritytech/substrate/pull/12891> and
                     <https://github.com/paritytech/substrate/pull/11381>.
   --> frame/assets/src/lib.rs:174:1
    |
174 | #[frame_support::pallet]
    | ^^^^^^^^^^^^^^^^^^^^^^^^
```
2.
```pre
warning: use of deprecated field `pallet::Call::transfer_approved::destination`: 
                     Implicit call indices are deprecate in favour of explicit ones.
                     Please ensure that all calls have the `pallet::call_index` attribute or that the
                     `dev-mode` of the pallet is enabled. For more info see:
                     <https://github.com/paritytech/substrate/pull/12891> and
                     <https://github.com/paritytech/substrate/pull/11381>.
    --> frame/assets/src/lib.rs:1418:4
     |
1418 |             destination: AccountIdLookupOf<T>,
     |             ^^^^^^^^^^^
```

Reply:
Ah okay, then let's continue to use new structs instead of the enum variants.